### PR TITLE
Reducing complexity of execute method in SuggestCommand

### PR DESCRIPTION
**Pull request by Idrinth**
I took the liberty of ripping appart the execute method to make other goals(like testing) easier to achieve.

The following was done:
- moved duplicate code to methods
- extracted classes to wrap logically related functionality or data

### CODE REVIEW

**Comment by Idrinth** [comment]
just wondering, should I reformat the whole thing to match psr2? would look cleaner obviously and could be tested automatically, on the other hand it might makie diffs across that change harder to read.

**Comment by tomzx** [comment]
Please keep the style as is for now. I'd rather not have style change because of an unrelated PR.

File: bin/php-semver-checker-git [hunk 1]
```diff
@@ -1,28 +1,18 @@
...
-foreach ($autoload_paths as $path) {
+foreach (array(
+             __DIR__.'/../vendor/autoload.php',
+             __DIR__.'/../../../autoload.php',
+         ) as $path) {
 	if (file_exists($path)) {
-		require $path;
-		$found = true;
-		break;
+		require($path);
```
Comment (tomzx) [inline bin/php-semver-checker-git]: This is a bit unrelated to this PR.

File: composer.json [hunk 2]
```diff
@@ -35,8 +35,7 @@
 		"vierbergenlars/php-semver": "^3.0"
 	},
 	"require-dev": {
-		"mockery/mockery": "^1.0",
-		"phpunit/phpunit": "^4.0|^5.0"
+		"phpunit/phpunit": "^4.8.36|^5.7.27"
```
Comment (tomzx) [inline composer.json]: Why remove mockery and not phpunit as well, given there are currently no tests?

File: src/PHPSemVerCheckerGit/Console/Command/SuggestCommand.php [hunk 3]
```diff
@@ -7,17 +7,17 @@
...
+use PHPSemVerCheckerGit\SourceFileProcessor;
 use RuntimeException;
-use Symfony\Component\Console\Helper\ProgressBar;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Input\InputOption;
 use Symfony\Component\Console\Output\OutputInterface;
 use vierbergenlars\SemVer\expression as SemanticExpression;
 use vierbergenlars\SemVer\SemVerException as SemanticVersionException;
 use vierbergenlars\SemVer\version as SemanticVersion;
+use PHPSemVerChecker\Report\Report;
```
Comment (tomzx) [inline src/PHPSemVerCheckerGit/Console/Command/SuggestCommand.php]: Please follow alphabetical ordering.

File: test/Console/Command/SuggestCommandTest.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            array(Level::NONE, $this->getMockedVersion(1,0,'0-1'), '1.0.0-1'),
+            array(Level::PATCH, $this->getMockedVersion(1,0,'0-1'), '1.0.0-2'),
+            array(Level::MINOR, $this->getMockedVersion(1,0,'0-1'), '1.0.0-2'),
+            array(Level::MAJOR, $this->getMockedVersion(1,0,'0-1'), '1.0.0-2'),
+            //v1.0.0
+            array(Level::NONE, $this->getMockedVersion(1,0,0), '1.0.0'),
+            array(Level::PATCH, $this->getMockedVersion(1,0,0), '1.0.1'),
+            array(Level::MINOR, $this->getMockedVersion(1,0,0), '1.1.0'),
+            array(Level::MAJOR, $this->getMockedVersion(1,0,0), '2.0.0'),
+            //v2.3.4
```
Comment (tomzx) [inline test/Console/Command/SuggestCommandTest.php]: Anything in particular not tested by the v1.0.0 case?

File: test/Console/Command/SuggestCommandTest.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * @param $level
+     * @param $version
+     * @param $expected
+     * @throws \ReflectionException
+     * @test
+     * @dataProvider provideGetNextTag
+     */
+    public function testGetNextTag($level, version $version, $expected) {
```
Comment (tomzx) [inline test/Console/Command/SuggestCommandTest.php]: This would probably be simpler to read using mockery.

File: test/Console/Command/SuggestCommandTest.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace PHPSemVerCheckerGit\Test;
+
+use PHPSemVerCheckerGit\Console\Command\SuggestCommand;
+use PHPUnit\Framework\TestCase;
+use PHPSemVerChecker\SemanticVersioning\Level;
+use vierbergenlars\SemVer\version;
+
+class SuggestCommandTest extends TestCase
```
Comment (tomzx) [inline test/Console/Command/SuggestCommandTest.php]: :+1: for the added tests.

File: src/PHPSemVerCheckerGit/ProcessedFileList.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    private $scanner;
+
+    /**
+     * ProcessedFileList constructor.
+     * @param int $originalAmount
+     * @param string[] $files
+     * @param Scanner $scanner
+     */
+    public function __construct($originalAmount, array &$files, Scanner &$scanner)
```
Comment (tomzx) [inline src/PHPSemVerCheckerGit/ProcessedFileList.php]: I'd rather have the list of original (unfiltered) files here than the count.

Comment (tomzx) [inline src/PHPSemVerCheckerGit/ProcessedFileList.php]: Why a reference to the files array?

File: src/PHPSemVerCheckerGit/SourceFileProcessor.php [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+
+    /**
+     * @param string $commitIdentifier
+     * @param $include
+     * @param $exclude
+     * @return ProcessedFileList
+     */
+    public function processFileList(
```
Comment (tomzx) [inline src/PHPSemVerCheckerGit/SourceFileProcessor.php]: Why is this on multiple lines?

File: src/PHPSemVerCheckerGit/SourceFileProcessor.php [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        $source = $this->filter->filter($source, $this->modifiedFiles);
+        $this->scanFileList($scanner, $source);
+        return new ProcessedFileList($count, $source, $scanner);
+    }
+
+    /**
+     * @param Scanner $scanner
+     * @param array $files
+     */
+    private function scanFileList(Scanner &$scanner, array &$files)
```
Comment (tomzx) [inline src/PHPSemVerCheckerGit/SourceFileProcessor.php]: Why references?

File: test/Console/Command/SuggestCommandTest.php (hunk 4 above)
Comment (Idrinth) [reply test/Console/Command/SuggestCommandTest.php]: making sure the incrementing does zero out the next lower version number. Might be a bit more than required.

File: bin/php-semver-checker-git (hunk 1 above)
Comment (Idrinth) [reply bin/php-semver-checker-git]: agreed, not quite sure why I committed this to this branch. Should I remove it?

Comment (tomzx) [reply bin/php-semver-checker-git]: Yes. Please keep the PR contained :wink:.

File: src/PHPSemVerCheckerGit/Console/Command/SuggestCommand.php [hunk 10]
```diff
@@ -50,8 +50,8 @@ private function getRepository($directory)
     }
 
 	/**
-	 * @param \Symfony\Component\Console\Input\InputInterface   $input
-	 * @param \Symfony\Component\Console\Output\OutputInterface $output
+	 * @param InputInterface   $input
```
Comment (tomzx) [inline src/PHPSemVerCheckerGit/Console/Command/SuggestCommand.php]: Be careful, we still want the FQCN (fully qualified class name).

File: src/PHPSemVerCheckerGit/ProcessedFileList.php [hunk 11]
```diff
@@ -40,20 +54,35 @@ public function getScanner()
...
+    public function getFiltered()
+    {
+        return $this->filtered;
+    }
+
     /**
      * @return int
      */
-    public function getOriginalAmount()
+    public function getFilteredAmount()
```
Comment (tomzx) [inline src/PHPSemVerCheckerGit/ProcessedFileList.php]: Use "count" instead of "amount" (see https://english.stackexchange.com/questions/141564/differences-between-amount-count-number-and-quantity)

File: src/PHPSemVerCheckerGit/Console/Command/SuggestCommand.php (hunk 10 above)
Comment (Idrinth) [reply src/PHPSemVerCheckerGit/Console/Command/SuggestCommand.php]: having FQN in PHPdoc only looks inconsistent to me when they aren't also used elsewhere, Will fix those.

File: src/PHPSemVerCheckerGit/ProcessedFileList.php (hunk 11 above)
Comment (Idrinth) [reply src/PHPSemVerCheckerGit/ProcessedFileList.php]: very useful, thank you.

**Comment by Idrinth** [comment]
is there any todo left @tomzx ?

**Comment by tomzx** [comment]
I'll review this again over the next few days. Thanks for the ping.

**Comment by Idrinth** [comment]
thanks :)
